src/linker/Linker/CompilerGeneratedState.cs

```diff
@@ -165,6 +185,77 @@ void ProcessMethod (MethodDefinition method)
 					}
 				}
 			}
+
+			// Now that we have instantiating methods fully filled out, walk the state machines and fill in the attribute
+			// providers
+			foreach (var stateMachineType in _generatedTypeToTypeArgumentInfo.Keys) {
+				if (!stateMachineType.HasGenericParameters) {
+					continue;
+				}
+				MapGeneratedTypeTypeParameters (stateMachineType);
+			}
+
+			return;
+
+			void MapGeneratedTypeTypeParameters (TypeDefinition stateMachineType)
+			{
+				var typeInfo = _generatedTypeToTypeArgumentInfo[stateMachineType];
+				if (typeInfo.OriginalAttributes is not null) {
+					return;
+				}
+				var method = typeInfo.CreatingMethod;
+				if (method.Body is { } body) {
+					var typeArgs = new ICustomAttributeProvider[stateMachineType.GenericParameters.Count];
+					var typeRef = ScanForInit (stateMachineType, body);
+					if (typeRef is null) {
+						return;
+					}
+					for (int i = 0; i < typeRef.GenericArguments.Count; i++) {
+						var typeArg = typeRef.GenericArguments[i];
+						// The type parameters of the state machine types are alpha renames of the
+						// the method parameters, so the type ref should always be a GenericParameter. However,
+						// in the case of nesting, there may be multiple renames, so if the parameter is a method
+						// we know we're done, but if it's another state machine, we have to keep looking to find
+						// the original owner of that state machine.
+						if (typeArg is GenericParameter param) {
+							if (param.Owner is MethodReference) {
+								typeArgs[i] = param;
+								continue;
+							} else {
+								var parentType = _context.TryResolve ((TypeReference) param.Owner);
+								if (parentType is not null) {
+									PopulateCacheForType (parentType.DeclaringType);
+									MapGeneratedTypeTypeParameters (parentType);
+									if (_generatedTypeToTypeArgumentInfo[parentType].OriginalAttributes is { } parentAttrs) {
+										typeArgs[i] = parentAttrs[param.Position];
+										continue;
+									}
+								}
+							}
+						}
+
+						// This should probably never happen in valid code
+						typeArgs[i] = stateMachineType.GenericParameters[i];
+					}
+					_generatedTypeToTypeArgumentInfo[stateMachineType] = typeInfo with { OriginalAttributes = typeArgs };
+				}
+			}
+
+			GenericInstanceType? ScanForInit (TypeDefinition stateMachineType, MethodBody body)
+			{
+				foreach (var instr in body.Instructions) {
+					switch (instr.OpCode.Code) {
+					case Code.Initobj:
+					case Code.Newobj:
+						if (instr.Operand is MethodReference { DeclaringType: GenericInstanceType typeRef }
+							&& stateMachineType.MetadataToken == _context.TryResolve (typeRef)?.MetadataToken) {
```

Comment:
I think you can compare the TypeDefinition by reference - I think Cecil guarantees uniqueness. Note that we don't use special comparers when using TypeDefinition as a key in a Dictionary.